addons/web_editor/tools.py

```diff
@@ -0,0 +1,133 @@
+
+# Regex for few of the widely used video hosting services
+player_regexes = {
+    'youtube': r'^(?:(?:https?:)?\/\/)?(?:www\.)?(?:youtu\.be\/|youtube(-nocookie)?\.com\/(?:embed\/|v\/|watch\?v=|watch\?.+&v=))((?:\w|-){11})\S*$',
+    'vimeo': r'\/\/(player.)?vimeo.com\/([a-z]*\/)*([0-9]{6,11})[?]?.*',
+    'dailymotion': r'(?:.+dailymotion.com\/(video|hub|embed\/video|embed)|dai\.ly)\/([^_?]+)[^#]*(#video=([^_&]+))?',
+    'instagram': r'(?:(.*)instagram.com|instagr\.am)\/p\/(.[a-zA-Z0-9-_\.]*)',
+    'youku': r'(.*).youku\.com\/(v_show\/id_|embed\/)(.+)',
+}
+
+
+def get_video_source_data(video_url):
+    ''' Computes the valid source, document ID and regex match from given URL
+        (or False in case of invalid URL).
+    '''
+    if not video_url:
+        return False
+
+    if re.search(valid_url_regex, video_url):
+        youtube_match = re.search(player_regexes['youtube'], video_url)
+        if youtube_match:
+            return ('youtube', youtube_match.groups()[1], youtube_match)
+        vimeo_match = re.search(player_regexes['vimeo'], video_url)
+        if vimeo_match:
+            return ('vimeo', vimeo_match.groups()[2], vimeo_match)
+        dailymotion_match = re.search(player_regexes['dailymotion'], video_url)
+        if dailymotion_match:
+            return ('dailymotion', dailymotion_match.groups()[1], dailymotion_match)
+        instagram_match = re.search(player_regexes['instagram'], video_url)
+        if instagram_match:
+            return ('instagram', instagram_match.groups()[1], instagram_match)
+        youku_match = re.search(player_regexes['youku'], video_url)
+        if youku_match:
+            youku_link = youku_match.groups()[2]
+            if '.html?' in youku_link:
+                youku_link = youku_link.split('.html?')[0]
+            return ('youku', youku_link, youku_match)
+    return False
+
+
+def get_video_url_data(video_url, **kwargs):
+    ''' Computes the platform name and embed_url from given URL
+        (or error message in case of invalid URL).
+    '''
+    source = get_video_source_data(video_url)
+    if not source:
+        return {'error': True, 'message': _('The provided url is invalid')}
+
+    autoplay = kwargs.get('autoplay') and '?autoplay=1&mute=1' or '?autoplay=0'
+    controls = kwargs.get('hide_controls') and '&controls=0' or ''
+    loop = kwargs.get('loop') and '&loop=1' or ''
+
+    # We directly use the provided URL as it is
+    embed_url = video_url
+    platform = source[0]
+    platform_id = source[1]
+    platform_match = source[2]
+    if platform == 'youtube':
+        fullscreen = kwargs.get('hide_fullscreen') and '&fs=0' or ''
+        youtube_loop = loop and loop + '&playlist=%s' % (platform_id) or ''
+        logo = kwargs.get('hide_yt_logo') and '&modestbranding=1' or ''
+        params = '%s&rel=0%s%s%s%s' % (autoplay, youtube_loop, controls, fullscreen, logo)
```

Comment:
Why not just build up a map of query parameters and use `url_encode` to generate the final query string?

Reply:
Because I was only merging a two years-old PR which is just moving code but sure :+1: